### ISSUE TITLE
[RFC] Deprecate the _epoch_end hooks

### ISSUE THREAD

**Issue by ananthsub** [PyTorchLightning/pytorch-lightning#8731]
Lightning has had some recent issues filed around these hooks:
- `training_epoch_end`
- `validation_epoch_end`
- `test_epoch_end`
- `predict_epoch_end`

Examples:
- https://github.com/PyTorchLightning/pytorch-lightning/issues/8603
- https://github.com/PyTorchLightning/pytorch-lightning/issues/8657
- https://github.com/PyTorchLightning/pytorch-lightning/issues/8670 
- https://github.com/PyTorchLightning/pytorch-lightning/issues/8479

Do we really need these on the LightningModule? Why don't users track these outputs directly inside the LightningModule?

Asking users to do this can offer some benefits:
- less API surface area for the trainer reduces risk of bugs like this. these bugs hurt user trust because the control flow isn't visible to end users, so removing this class of bugs can have a disproportionate benefit to user trust
- the trainer doesn't take responsibility of data it doesn't directly use. this simplifies the loop internals
- less "framework magic" means more readable user code because this tracking is explicit
- because the tracking is explicit, the responsibility of testing also falls to users
- we avoid API confusion: for instance, when should users implement something in `training_epoch_end` vs `on_train_epoch_end` ? this can improve the onboarding experience (one less class of hooks to learn about, only 1 way to do things). 
- it can also improve performance: if users implement something in `training_epoch_end` and don't use `outputs`, the trainer accumulates results needlessly which wastes memory. This is not clearly visible to the user either.

Cons:
- (marginally) more boilerplate code in LightningModules. for instance, users would need to pay attention to resetting the results in `on_{train/val}_epoch_start` unless they want to accumulate results across epochs. 

@PyTorchLightning/core-contributors

_Originally posted by @ananthsub in https://github.com/PyTorchLightning/pytorch-lightning/discussions/8690_

**Comment by ananthsub** [comment]
one thing that will emerge is the hook call order between callbacks and the lightning module. logging something in the module's `on_train_epoch_end`  needs to happen first for it to be used in the callback's `on_train_epoch_end` hook. this would be a change compared to today's behavior. this is required for items like logging a value at the end of the epoch in the module, and using that key/value as a monitor for callbacks like model checkpointing or early stopping.

so this will be related to https://github.com/PyTorchLightning/pytorch-lightning/issues/8506

**Comment by carmocca** [comment]
Even though I agree with your arguments, people just love "framework magic". Removing this can make sense engineering-wise and design-wise but it can break user trust.

> This is how easily users can implement this in their LightningModule with the existing hooks:

This does look simple but does not consider the cases where multiple optimizers are available. Basically, all the code we would remove would need to be implemented by the users who need it, depending on how complex is their training step.

If we were in a 0.x version I'd entirely agree with you, but we have to consider how used these hooks are and the degree of headaches that this change might impose on users

--- 

Have you considered making `output` aggregation opt-in? Requiring a LightningModule flag (similar to `automatic_optimization`).

If `outputs=True`, then the use of `training_epoch_end` is the same as before. If `outputs=False`, then using this hook is a way of dealing with the callback-model order.

**Comment by ananthsub** [comment]
> This does look simple but does not consider the cases where multiple optimizers are available. Basically, all the code we would remove would need to be implemented by the users who need it, depending on how complex is their training step.

I think that's also the point. If we want to support more flavors of training steps, the logic for handling these outputs in the framework gets more and more complex, when it could sit locally in the user's training step.

tracking the outputs per optimizer idx and dataloader idx in automatic optimization also shouldn't be significantly more work given these are accessible from the training step

**Comment by awaelchli** [comment]
API confusion aside, is the memory problem really an issue? I thought we are not tracking outputs if the hook is not overridden. In that sense, we already have the opt-in choice. This is the part where I don't get the full argument for removing it.

Regardless, I believe it would be interesting to see a lightning module rnn manual-implementation without the built-in TBTT. If we have that, we can examine the amount of boilerplate required and get a better picture of what impact this deprecation has.  Perhaps I could help here. I could try to contribute an example here.

**Comment by ananthsub** [comment]
> API confusion aside, is the memory problem really an issue? I thought we are not tracking outputs if the hook is not overridden. In that sense, we already have the opt-in choice. This is the part where I don't get the full argument for removing it.

@awaelchi The memory issue is the biggest risk. The current hook order makes it very dangerous:
- module.training_epoch_end
- callback.on_train_epoch_end
- module.on_train_epoch_end

if the user needs to log a value that's used in the callback's on train epoch end, then they currently are forced to do so in training_epoch_end (assuming they can't use training_step with self.log + on_epoch=True). and right away this incurs a major  performance hit (at best) or training failure (because of OOMs) 

I believe your second comment is meant for https://github.com/PyTorchLightning/pytorch-lightning/issues/8732

**Comment by carmocca** [comment]
> if the user needs to log a value that's used in the callback's on train epoch end, then they currently are forced to do so in training_epoch_end

After the logger connector re-design, if you have 2 callbacks `[A, B]` and `B` reads a value off `callback_metrics` logged in `A`, it works even if they both do it `on_train_epoch_end`. This is because the `callback_metrics` get updated dynamically:

https://github.com/PyTorchLightning/pytorch-lightning/blob/69f287eb85c36412d5d4d6541bc25f6a75a977ea/pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py#L297

So _technically_ you are not forced to implement `training_epoch_end` if your callbacks have a proper order

**Comment by anhnht3** [comment]
Would this affect logging of torch_metrics, i.e., automatically calling metric.compute() at the end of each epoch?

**Comment by ananthsub** [comment]
@carmocca the scenario I imagine is the LightningModule logs value that's required for callbacks `[A, B]` to process in `on_train_epoch_end`, not callbacks themselves logging metrics themselves for use in other callbacks. Though it's nice that the new logging supports this, I'm not sure users should need to rely on the order of callback execution

**Comment by daniellepintz** [comment]
What is the status of this issue? I saw it was added to a sprint in September

**Comment by ananthsub** [comment]
Another issue from transforming the return data from steps in a way users don't expect: https://github.com/PyTorchLightning/pytorch-lightning/issues/9968

**Comment by ananthsub** [comment]
Now that Lightning supports infinite training, these hooks introduce greater risks of OOMs: #11554
